zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs

```diff
@@ -0,0 +1,556 @@
+//! Add support for general transparent gap limit handling, and unify the `addresses` and
+//! `ephemeral_addresses` tables.
+
+use std::collections::HashSet;
+use uuid::Uuid;
+
+use rusqlite::{named_params, Transaction};
+use schemerz_rusqlite::RusqliteMigration;
+
+use zcash_keys::keys::UnifiedIncomingViewingKey;
+use zcash_protocol::consensus::{self, BlockHeight};
+
+use super::add_account_uuids;
+use crate::{
+    wallet::{self, init::WalletMigrationError, KeyScope},
+    AccountRef,
+};
+
+#[cfg(feature = "transparent-inputs")]
+use {
+    crate::{
+        wallet::{
+            decode_diversifier_index_be, encode_diversifier_index_be,
+            transparent::generate_gap_addresses,
+        },
+        GapLimits,
+    },
+    ::transparent::keys::{IncomingViewingKey as _, NonHardenedChildIndex},
+    zcash_keys::encoding::AddressCodec as _,
+    zcash_primitives::transaction::builder::DEFAULT_TX_EXPIRY_DELTA,
+    zip32::DiversifierIndex,
+};
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0xc41dfc0e_e870_4859_be47_d2f572f5ca73);
+
+const DEPENDENCIES: &[Uuid] = &[add_account_uuids::MIGRATION_ID];
+
+pub(super) struct Migration<P> {
+    pub(super) params: P,
+}
+
+impl<P> schemerz::Migration<Uuid> for Migration<P> {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        DEPENDENCIES.iter().copied().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Add support for general transparent gap limit handling, unifying the `addresses` and `ephemeral_addresses` tables."
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, conn: &Transaction) -> Result<(), WalletMigrationError> {
+        let decode_uivk = |uivk_str: String| {
+            UnifiedIncomingViewingKey::decode(&self.params, &uivk_str).map_err(|e| {
+                WalletMigrationError::CorruptedData(format!(
+                    "Invalid UIVK encoding {}: {}",
+                    uivk_str, e
+                ))
+            })
+        };
+
+        let external_scope_code = KeyScope::EXTERNAL.encode();
+
+        conn.execute_batch(&format!(
+            r#"
+            ALTER TABLE addresses ADD COLUMN key_scope INTEGER NOT NULL DEFAULT {external_scope_code};
+            ALTER TABLE addresses ADD COLUMN transparent_child_index INTEGER;
```

Comment:
An `exposed_at_height` column is not added here, so AFAICT that column will be `null` for every existing address in the table, while ephemeral addresses that get inserted have this set correctly. Given that going forward we set this for all address scopes (not just ephemeral AFAIR), we should set this correctly for existing addresses as well (to the best of our knowledge, i.e. no worse than if the wallet were recovered from seed).

Reply:
It should be safe here to set the exposed-at height to the wallet birthday; that's the most conservative approach.

Reply:
Fixed in c3e9b06fa3eef401aacafdca1dae609829bfe1d6